### DISCUSSION TITLE
Suggesting possible changes in Solveset's output for equations involving elementary functions like ln, exp, sin, cos

### DISCUSSION THREAD

**Discussion by anutosh491** [sympy/sympy#22406] [Ideas]
Hello guys, So I have been trying to approach couple of issues related to `Order.subs()` from the series module . I see that there are 2-3 issues reported on the same error . It arises in the following code block in `_eval_subs()`
```
sol = solveset(old - new.subs(var, d), d)
if isinstance(sol, Complement):
             e1 = sol.args[0]
             e2 = sol.args[1]
             sol = set(e1) - set(e2)
res = [dict(zip((d, ), sol))]
```
The problem is that `res` excepts `sol` to be a point but `sol` often gets a `ConditionSet` output for some equations . Now `ConditionSet` as output is not wrong but not what we expect in this case . These  mainly occur in cases involving `exp`, `ln`, `sin`, `cos`.
Consider the following examples
```
>>> O(x).subs(x, cos(x))
>>> O(x).subs(x, exp(1/x))
>>> O(x).subs(x, exp(x**2)) # All give back the same error 
  File "c:\users\anuto\sympy\sympy\sympy\series\order.py", line 478, in _eval_subs
    res = [dict(zip((d, ), sol))]
TypeError: 'ConditionSet' object is not iterable

```
This is because we don't get sol as a point ( I guess sol would take the real solution as point) and rather as a `ConditionSet` . 
For eg 
```
>>> solveset(x - exp(x), x, S.Reals) # Should return Empty set 
ConditionSet(x, Eq(x - exp(x), 0), Reals)

>>> solveset(x - sin(x), x, S.Reals) # Should return 0 
ConditionSet(x, Eq(x - sin(x), 0), Reals)

>>> solveset(x - cos(x), x, S.Reals)
ConditionSet(x, Eq(x - cos(x), 0), Reals)

>>> solveset(x - exp(1/x), x, S.Reals)
ConditionSet(x, Eq(x - exp(1/x), 0), Reals)

>>> solveset(x - ln(x), x, S.Reals) # Should return Empty Set
ConditionSet(x, Eq(x - log(x), 0), Reals)
```
Now as sympy has the `LambertW` function, we can solve many of these including `trig` functions if expressed as `exp` as done in other CAS libraries etc like Wolfram, Diofant. ( links to results attached below )
https://www.wolframalpha.com/input/?i2d=true&i=x+%3D+Power%5Be%2CDivide%5B1%2Cx%5D%5D
https://www.wolframalpha.com/input/?i2d=true&i=x+%3D+Power%5Be%2Cx%5D

Hence I suggest that we should try to return solutions for these in terms of `LambertW` ( and possibly as a point) if possible and if not then `ConditionSet `should be used ! I am completely new to the Solveset Module so not sure whether this is a good suggestion or not but would surely help in addressing some 2-3 issues from the series module .

**Comment by oscarbenjamin** [comment]
A function like `solveset` is far to heavy to be called as part of `subs`. That code should be changed I think.

**Reply by anutosh491** [reply]
Yeah it makes sense! even that should be changed

**Reply by anutosh491** [reply]
What is your take on solveset returning points using LambertW which is actually what one would expect rather than using ConditionSet ? We could try to use LambertW and if we don't succed we could go for ConditionSet

**Reply by oscarbenjamin** [reply]
Yes, it should use LambertW. There were PRs attempting to add that before.

**Reply by anutosh491** [reply]
> Yes, it should use LambertW. There were PRs attempting to add that before.

Ohh ok ,I wasn't aware of this , will look into it and maybe try to find one of these .

Anyways, as you said we should not be using solveset as a part of subs ...... do you have any suggestions how we should go about fixing this . As I am relatively new to the Solvers Module ,I would be glad to have any suggestions/thoughts from your side ! @oscarbenjamin
